src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs

```diff
@@ -259,7 +234,78 @@ IEnumerator IEnumerable.GetEnumerator()
             }
         }
 
-        private class ConfigurationBuilderProperties : IDictionary<string, object>
+        private sealed class ProviderManager : IDisposable
+        {
+            private readonly object _replaceProvidersLock = new object();
+            private ProvidersReference _refCountedProviders = new(new List<IConfigurationProvider>());
+
+            public IEnumerable<IConfigurationProvider> Providers => _refCountedProviders.Providers;
+
+            public ProvidersReference GetReference()
+            {
+                // Lock to ensure oldRefCountedProviders.Dispose() in ReplaceProviders() doesn't decrement ref count to zero
+                // before calling _refCountedProviders.AddRef().
+                lock (_replaceProvidersLock)
+                {
+                    _refCountedProviders.AddRef();
+                    return _refCountedProviders;
+                }
+            }
+
+            // Providers should never be concurrently modified. Reading during modification is allowed.
+            public void ReplaceProviders(List<IConfigurationProvider> providers)
+            {
+                ProvidersReference oldRefCountedProviders = _refCountedProviders;
+
+                lock (_replaceProvidersLock)
+                {
+                    _refCountedProviders = new ProvidersReference(providers);
+                }
+
+                oldRefCountedProviders.Dispose();
+            }
+
+            public void AddProvider(IConfigurationProvider provider)
+            {
+                // Maintain existing references, but replace list with copy containing new item.
+                _refCountedProviders.Providers = new List<IConfigurationProvider>(_refCountedProviders.Providers)
```

Comment:
Technically this should be done with a volatile write, in order to ensure all of the writes that initialize the list are published before the write that updates Providers.  In practice, the runtime needs a barrier anyway, so it won't matter.  I personally prefer to make it clear in the code.